mixnet/client/src/sender.rs

```diff
@@ -126,28 +126,58 @@ impl<R> Iterator for RandomDelayIterator<R>
 where
     R: Rng,
 {
-    type Item = u64;
+    type Item = Duration;
 
-    fn next(&mut self) -> Option<u64> {
+    fn next(&mut self) -> Option<Duration> {
         if self.remaining_delays == 0 {
             return None;
         }
 
+        self.remaining_delays -= 1;
+
         if self.remaining_delays == 1 {
-            self.remaining_delays -= 1;
-            return Some(self.remaining_time);
+            return Some(Duration::from_millis(self.remaining_time));
         }
 
         // Calculate bounds to avoid extreme values
-        let lower_bound = (self.avg_delay as f64 * 0.5) as u64;
         let upper_bound = (self.avg_delay as f64 * 1.5)
-            .min((self.remaining_time - self.remaining_delays + 1) as f64)
-            as u64;
+            // guarantee that we don't exceed the remaining time and promise the delay we return is
+            // at least 1ms.
+            .min(self.remaining_time.saturating_sub(self.remaining_delays) as f64);
+        let lower_bound = (self.avg_delay as f64 * 0.5).min(upper_bound);
 
-        let delay = self.rng.gen_range(lower_bound, upper_bound);
-        self.remaining_time -= delay;
-        self.remaining_delays -= 1;
+        let delay = Uniform::new_inclusive(lower_bound, upper_bound).sample(&mut self.rng) as u64;
+        self.remaining_time = self.remaining_time.saturating_sub(delay);
+
+        Some(Duration::from_millis(delay))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::time::Duration;
 
-        Some(delay)
+    use super::RandomDelayIterator;
+
+    const TOTAL_DELAYS: u64 = 3;
+
+    #[test]
+    fn test_random_delay_iter_zero_total_time() {
+        let mut delays = RandomDelayIterator::new(rand::thread_rng(), TOTAL_DELAYS, 0);
+        for _ in 0..TOTAL_DELAYS {
+            assert!(delays.next().is_some());
+        }
+        assert!(delays.next().is_none());
+    }
+
+    #[test]
+    fn test_random_delay_iter_small_total_time() {
+        let mut delays = RandomDelayIterator::new(rand::thread_rng(), TOTAL_DELAYS, 1);
```

Comment:
in `RandomDelayIterator::new(rng, total_delays, total_time)`, `total_time` should be a `Duration` and not a `u64`